### PR TITLE
Added JApplicationInterface

### PR DESCRIPTION
**Pull request by robschley**
Added JApplicationInterface to provide a starting point for a common application interface and to support better type hinting. 
I also made JWeb, JCli, and JDaemon all support the new interface.

### CODE REVIEW

File: libraries/joomla/application/cli.php [hunk 1]
```diff
@@ -325,6 +331,58 @@ public function set($key, $value = null)
...
+	 * @param   string  $charset  The character set.
+	 *
+	 * @return  void
+	 *
+	 * @since   12.1
+	 */
+	public function setCharacterSet($charset)
+	{
+		$this->charSet = $charset;
+	}
```
Comment (chdemko) [inline libraries/joomla/application/cli.php]: Could we return $this to allow chaining?

File: libraries/joomla/application/cli.php [hunk 2]
```diff
@@ -325,6 +331,58 @@ public function set($key, $value = null)
...
+	 * @param   JInput  $input  The input object.
+	 *
+	 * @return  void
+	 *
+	 * @since   12.1
+	 */
+	public function setInput(JInput $input)
+	{
+		$this->input = $input;
+	}
```
Comment (chdemko) [inline libraries/joomla/application/cli.php]: Could we return $this to allow chaining?

File: libraries/joomla/application/web.php [hunk 3]
```diff
@@ -712,6 +712,58 @@ public function set($key, $value = null)
...
+	 * @param   string  $charset  The character set.
+	 *
+	 * @return  void
+	 *
+	 * @since   12.1
+	 */
+	public function setCharacterSet($charset)
+	{
+		$this->charSet = $charset;
+	}
```
Comment (chdemko) [inline libraries/joomla/application/web.php]: Could we return $this to allow chaining?

File: libraries/joomla/application/web.php [hunk 4]
```diff
@@ -712,6 +712,58 @@ public function set($key, $value = null)
...
+	 * @param   JInput  $input  The input object.
+	 *
+	 * @return  void
+	 *
+	 * @since   12.1
+	 */
+	public function setInput(JInput $input)
+	{
+		$this->input = $input;
+	}
```
Comment (chdemko) [inline libraries/joomla/application/web.php]: Could we return $this to allow chaining?

File: libraries/joomla/application/interface.php [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	 */
+	public function getCharacterSet();
+
+	/**
+	 * Method to set the application character set.
+	 *
+	 * @param   string  $charset  The character set.
+	 *
+	 * @return  void
+	 *
```
Comment (chdemko) [inline libraries/joomla/application/interface.php]: Could we return $this to allow chaining?

File: libraries/joomla/application/interface.php [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	 * @since   12.1
+	 */
+	public function getInput();
+
+	/**
+	 * Method to set the application input.
+	 *
+	 * @param   JInput  $input  The input object.
+	 *
+	 * @return  void
```
Comment (chdemko) [inline libraries/joomla/application/interface.php]: Could we return $this to allow chaining?

**Comment by realityking** [comment]
Woult it make sense to start with a baseline that is supported by JApplication, JApplicationCli and JApplicationWeb?

**Pull request closed without merging**